standardtracer/tracer.go

```diff
@@ -2,16 +2,67 @@ package standardtracer
 
 import (
 	"sync"
+	"sync/atomic"
 	"time"
 
 	opentracing "github.com/opentracing/opentracing-go"
 )
 
-// New creates and returns a standard Tracer which defers to `recorder` after
-// RawSpans have been assembled.
-func New(recorder SpanRecorder) opentracing.Tracer {
+type sampleFunc func(int64) bool
+
+// Options allows creating a customized Tracer via NewWithOptions. The object
+// is thread safe and can be updated freely to influence the behavior of the
+// Tracer.
+type Options struct {
+	shouldSample       atomic.Value // contains a sampleFunc
+	trimUnsampledSpans int32        // updated atomically
+	recorder           atomic.Value // contains a SpanRecorder
```

Comment:
Adjusting the sampling rate in a running system, and we have a lot of testing situations in which it is very inconvenient to keep the system from moving until the test wants to set its tracing options (often, that would be a benchmark not keen on tracing). I'm not too worried about the performance overhead. Most of it is off the critical path (`trimUnsampledSpans` probably gets the most traffic), and atomic updates are pretty fast.
What pattern were you suggesting regarding immutability - I didn't quite follow.